test/unit/functions/noSelfOption.spec.ts

```diff
@@ -0,0 +1,52 @@
+import * as util from "../../util";
+
+test.each([["foo: string, bar: string", `this: void, foo: string, bar: string`]])(
+    'enables noSelfInFile behaviour for functions ("%s" equivalent to "%s")',
+    (expectedParameters, assignmentParameters) => {
+        util.testFunction`
+            function fooBar(${expectedParameters}) {}
+
+            const test: (${assignmentParameters}) => void = fooBar;
+        `
+            .setOptions({ noSelf: true })
+            .expectToHaveNoDiagnostics();
+    }
+);
+
+test.each([["foo: string, bar: string", `this: any, foo: string, bar: string`]])(
+    'enables noSelfInFile behaviour for methods ("%s" equivalent to "%s")',
+    (expectedParameters, assignmentParameters) => {
+        util.testFunction`
+            class FooBar {
+                fooBar(${expectedParameters}) {
+                    return foo + bar;
+                }
+            }
+            const fooBar = new FooBar();
+
+            const test: (${assignmentParameters}) => void = fooBar.fooBar;
+        `
+            .setOptions({ noSelf: true })
+            .expectToHaveNoDiagnostics();
+    }
+);
+
+test("generates declaration files with @noSelfInFile", () => {
+    const result = util.transpileStringsAsProject(
+        {
+            "main.ts": `function fooBar(foo: string, bar: string) {
+                return foo + bar;
+            }`,
+        },
+        {
+            noSelf: true,
+            declaration: true,
+        }
+    );
+
+    const declarationFile = result.transpiledFiles.find(transpiledFile => transpiledFile.declaration);
+
+    if (util.expectToBeDefined(declarationFile) && util.expectToBeDefined(declarationFile.declaration)) {
+        expect(declarationFile.declaration).toMatch(/^\/\*\* \@noSelfInFile \*\//);
```

Comment:
Actually, just checking string might be not enough to ensure that it works (for example now it doesn't fail, but we have `;` because of empty statement), it would be nicer to test it like that:
- Build a project with `noSelf: true`
- Take `declarationFile` from it's result
- Build a second project without `noSelf: true` and with a .d.ts file from previous project
- Test assignability like in previous tests

Something like that (haven't tested it though):

```ts
const builder = util.testModule`
    export function bar() {}
`
    .setOptions({ declaration: true, noSelf: true })
    .expectToHaveNoDiagnostics();

const declarationFile = builder.getLuaResult().transpiledFiles.find(f => f.declaration);
if (!util.expectToBeDefined(declarationFile) || !util.expectToBeDefined(declarationFile.declaration)) return;

util.testModule`
    import { bar } from "./foo";
    const test: (this: void) => void = bar;
`
    .addExtraFile("foo.d.ts", declarationFile.declaration)
    .expectToHaveNoDiagnostics();

```

Reply:
I've been trying to do this but I think I found a bug with with `util.testModule`.

```ts
test("test", () => {
    util.testModule`
        import { bar } from "./foo";
        bar();
    `
        .addExtraFile("foo.d.ts", "export function bar(): void;")
        .expectToHaveNoDiagnostics();
});
```

```
    main.ts:2:29 - error TS2307: Cannot find module './foo'.

    2         import { bar } from "./foo";
```

When renaming `"foo.d.ts"` to `"foo.ts"` and changing the contents of its file to `"export function foo() {}` this works but specifically declarations don't seem to work

Reply:
Resolved by using `import { bar } from "./foo.d";`